### PR TITLE
 Optimize bundling wheel by using multiprocessing in build_helper.sh

### PR DESCRIPTION
**Pull request by JanuszL**
- parallelize bundle-wheel for debug and regular wheel
- includes striping debugging symbol inside bundle-wheel to save
  one unzip->zip roundtrip
- moves test_bundled_libs.py to bundle-wheel to avoid another unzip call
- calls setup.py of DALI wheel directly to avoid copying data to tmp dir during creation
- changes compression method to `store` when setup.py is called as wheel is repacked again anyway
- reduces number of `patchelf --print-needed` calls to one per binary

Signed-off-by: Janusz Lisiecki <jlisiecki@nvidia.com>

## Description
- [ ] **Bug fix** (*non-breaking change which fixes an issue*)
- [ ] **New feature** (*non-breaking change which adds functionality*)
- [ ] **Breaking change** (*fix or feature that would cause existing functionality to not work as expected*)
- [x] **Refactoring** (*Redesign of existing code that doesn't affect functionality*)
- [ ] **Other** (*e.g. Documentation, Tests, Configuration*)

### What happened in this PR
- parallelize bundle-wheel for debug and regular wheel
- includes striping debugging symbol inside bundle-wheel to save
  one unzip->zip roundtrip
- moves test_bundled_libs.py to bundle-wheel to avoid another unzip call
- calls setup.py of DALI wheel directly to avoid copying data to tmp dir during creation
- changes compression method to `store` when setup.py is called as wheel is repacked again anyway
- reduces number of `patchelf --print-needed` calls to one per binary

#### Additional information
- Affected modules and functionalities:
  - bundle-wheel.sh
  - build_helper.sh

- Key points relevant for the review:
  - NA

### Tests
- [x] Existing tests apply
- [ ] New tests added
  - [ ] Python tests
  - [ ] GTests
  - [ ] Benchmark
  - [ ] Other
- [ ] N/A

### Documentation
- [ ] Existing documentation applies
- [ ] Documentation updated
  - [ ] Docstring
  - [ ] Doxygen
  - [ ] RST
  - [ ] Jupyter
  - [ ] Other
- [x] N/A

#### Requirements
- [ ] Implements new requirements
- [ ] Affects existing requirements
- [x] N/A

**REQ IDs**: N/A

**JIRA TASK**: DALI-2280

### CODE REVIEW

**Comment by JanuszL** [comment]
!build

**Comment by JanuszL** [comment]
!build

**Comment by JanuszL** [comment]
!build

**Comment by JanuszL** [comment]
It should save 2-3 minutes according to CI log.

**Comment by JanuszL** [comment]
!build

**Comment by JanuszL** [comment]
!build

File: docker/build_helper.sh [hunk 1]
```diff
@@ -134,7 +126,18 @@ if [ "${BUILD_PYTHON}" = "ON" ]; then
         ### `--set-interpreter' with a larger path than the original is used).
         ### This appears to be a bug in binutils
         ### (http://bugs.strategoxt.org/browse/NIXPKGS-85).
+        bundle_wheel nvidia_dali[_-]*.whl NO ${OUT_DEBUG_WHL_NAME} "${BUNDLE_PATH_PREFIX}" &
+        bundle_wheel nvidia_dali[_-]*.whl YES ${OUT_WHL_NAME} "${BUNDLE_PATH_PREFIX}" &
+        wait
+    else
+        bundle_wheel nvidia_dali[_-]*.whl NO ${OUT_WHL_NAME} "${BUNDLE_PATH_PREFIX}" &
+        wait
```
Comment (mzient) [inline docker/build_helper.sh:134]: Nitpick: it could just be synchronous?

File: docker/build_helper.sh [hunk 2]
```diff
@@ -134,7 +126,18 @@ if [ "${BUILD_PYTHON}" = "ON" ]; then
...
+    fi
 
-        ../dali/python/bundle-wheel.sh nvidia_dali[_-]*.whl  ${BUNDLE_PATH_PREFIX}
+    if [ "${TEST_BUNDLED_LIBS}" = "YES" ]; then
+        export UNZIP_PATH="$(mktemp -d)"
+        unzip /wheelhouse/$(ls -I '*_debug.whl' /wheelhouse/) -d $UNZIP_PATH
+        python ../tools/test_bundled_libs.py $(find $UNZIP_PATH -iname *.so* | tr '\n' ' ')
```
Comment (mzient) [inline docker/build_helper.sh:140]: How long does it take? Could this be run in parallel? Either here or in `test_bundled_libs.py` (I think that even the gelded Python threads could be of use when launching an external program to do the work).

Comment (JanuszL) [reply docker/build_helper.sh:140]: It is ~20 whole, and 95% of it is taken by `unzip`.

Comment (JanuszL) [reply docker/build_helper.sh:140]: I can make it part of bundle-wheel.sh.

Comment (JanuszL) [reply docker/build_helper.sh]: Moved.

File: docker/build_helper.sh (hunk 1 above)
Comment (JanuszL) [reply docker/build_helper.sh]: Fixed

**Comment by JanuszL** [comment]
!build

**Comment by JanuszL** [comment]
Now we saved ~300s compared to what we used to have.

File: dali/python/bundle-wheel.sh [hunk 3]
```diff
@@ -116,6 +119,19 @@ unzip -q $INWHL
 mkdir -p $PKGNAME_PATH/.libs
 popd
 
+stip_so () {
```
Comment (mzient) [inline dali/python/bundle-wheel.sh:122]:
```suggestion
strip_so () {
```
?

File: dali/python/bundle-wheel.sh [hunk 4]
```diff
@@ -116,6 +119,19 @@ unzip -q $INWHL
...
+    local filepath=$1
+    strip --strip-debug $filepath
+}
+
+if [[ "$STRIP_DEBUG" != "NO" ]]; then
+    echo "Striping .so files from debug info"
+    for f in $(find $TMPDIR -iname *.so); do
+        stip_so $f &
```
Comment (mzient) [inline dali/python/bundle-wheel.sh:130]:
```suggestion
        strip_so $f &
```

File: dali/python/bundle-wheel.sh (hunk 3 above)
Comment (JanuszL) [reply dali/python/bundle-wheel.sh]: Done

File: dali/python/bundle-wheel.sh (hunk 4 above)
Comment (JanuszL) [reply dali/python/bundle-wheel.sh]: Done

**Comment by JanuszL** [comment]
!build

**Review by mzient — APPROVED** [review]

**Comment by JanuszL** [comment]
!build

**Comment by JanuszL** [comment]
!build

**Review by mzient — APPROVED** [review]

**Review by klecki — APPROVED** [review]

**Comment by JanuszL** [comment]
!build

**Pull request merged**